### PR TITLE
[Session] Deprecate session configurators in favor of event listeners

### PR DESCRIPTION
**Pull request by dvesh3**
## Changes in this pull request  
Related to #12241

### CODE REVIEW

File: bundles/AdminBundle/EventListener/AdminSessionBagListener.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Twig\Environment;
+
+/**
+ * @internal
+ */
+class AdminSessionBagListener implements EventSubscriberInterface
+{
+    public function __construct(protected RequestStack $requestStack, protected Config $config)
```
Comment (jdreesen) [inline bundles/AdminBundle/EventListener/AdminSessionBagListener.php]: The `RequestStack` seems to be unused.

```suggestion
    public function __construct(protected Config $config)
```

Comment (jdreesen) [inline bundles/AdminBundle/EventListener/AdminSessionBagListener.php:35]: The `RequestStack` seems to be unused.

```suggestion
    public function __construct(protected Config $config)
```

File: doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md [hunk 2]
```diff
@@ -5,8 +5,13 @@
   Please note, if the routing import config is in use, it is recommended to correct the config path (by removing `/Resources`) to follow the [new folder tree structure](https://github.com/prestaconcept/PrestaSitemapBundle/releases/tag/v3.0.0),
   eg. "@PrestaSitemapBundle/~~Resources/~~config/routing.yaml", to ensure a smoother upgrade to upcoming major release.
 - [Backend search] `key` and `index` columns have been added to the search index. Run `./bin/console pimcore:search-backend-reindex` to reindex.
+- [Session] Implementing Session Configurator with tag `pimcore.session.configurator` to register session bags, is deprecated and will be removed in Pimcore 11.
+  Implement an (EventListener)[] to register a session bag before the session is started.
```
Comment (jdreesen) [inline doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md]: This seems to be an incomplete link.

Comment (jdreesen) [inline doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md:9]: This seems to be an incomplete link.

File: bundles/AdminBundle/EventListener/AdminSessionBagListener.php (hunk 1 above)
Comment (dvesh3) [reply bundles/AdminBundle/EventListener/AdminSessionBagListener.php]: Removed it. thanks!

File: doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md (hunk 2 above)
Comment (dvesh3) [reply doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md]: Added. thanks!

File: bundles/EcommerceFrameworkBundle/PricingManager/Condition/Token.php [hunk 3]
```diff
@@ -18,6 +18,9 @@
 use Pimcore\Bundle\EcommerceFrameworkBundle\PricingManager\ConditionInterface;
 use Pimcore\Bundle\EcommerceFrameworkBundle\PricingManager\EnvironmentInterface;
 
+/**
+ * @deprecated will be removed in Pimcore 11
+ */
```
Comment (fashxp) [inline bundles/EcommerceFrameworkBundle/PricingManager/Condition/Token.php:23]: we should trigger deprecation warnings here...

File: bundles/EcommerceFrameworkBundle/Resources/config/services.yaml [hunk 4]
```diff
@@ -10,8 +10,10 @@ services:
 
     Pimcore\Bundle\EcommerceFrameworkBundle\Tools\SessionConfigurator:
         public: false
-        tags:
-            - { name: pimcore.session.configurator }
+        deprecated:
```
Comment (fashxp) [inline bundles/EcommerceFrameworkBundle/Resources/config/services.yaml]: why are the tags removed here?

File: lib/Targeting/Session/SessionConfigurator.php [hunk 5]
```diff
@@ -17,79 +17,11 @@
...
-use Symfony\Component\HttpFoundation\Cookie;
-use Symfony\Component\HttpFoundation\Session\Attribute\AttributeBag;
-use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Pimcore\Targeting\EventListener\TargetingSessionBagListener;
 
-class SessionConfigurator implements SessionConfiguratorInterface, EventSubscriberInterface
+/**
+ * @deprecated will be removed in Pimcore 11. Use TargetingSessionBagListener instead.
+ */
+class SessionConfigurator extends TargetingSessionBagListener
```
Comment (fashxp) [inline lib/Targeting/Session/SessionConfigurator.php]: can we really move `configureIgnoredSessionKeys` and `prepareFullPageCacheResponse` already? 
that could be a BC break if someone has overwritten them to change behavior.

Comment (fashxp) [reply lib/Targeting/Session/SessionConfigurator.php]: I'm not sure if we even can move `configure` already?

File: bundles/EcommerceFrameworkBundle/Tools/SessionConfigurator.php [hunk 6]
```diff
@@ -15,60 +15,12 @@
...
-        ];
-    }
-
-    /**
-     * {@inheritdoc}
-     */
-    public function configure(SessionInterface $session)
-    {
-        $bagNames = $this->getBagNames();
+use Pimcore\Bundle\EcommerceFrameworkBundle\EventListener\SessionBagListener;
```
Comment (fashxp) [inline bundles/EcommerceFrameworkBundle/Tools/SessionConfigurator.php:19]: we definitely cannot move the constants already. we also cannot remove the `clearSession` method already.

File: tests/Model/Permissions/ModelAssetPermissionsTest.php [hunk 7]
```diff
@@ -725,23 +725,23 @@ public function testManyElementSearch()
         //update role
         $role = User\Role::getByName('Testrole');
         $role->setWorkspacesAsset([
-            (new User\Workspace\Asset())->setValues(['cId' => $manyElementX->getId(), 'cPath' => $manyElementX->getFullpath(), 'list' => true, 'view' => true]),
-            (new User\Workspace\Asset())->setValues(['cId' => $this->groupfolder->getId(), 'cPath' => $this->groupfolder->getFullpath(), 'list' => true, 'view' => true]),
+            (new User\Workspace\Asset())->setValues(['cId' => $manyElementX->getId(), 'cPath' => $manyElementX->getRealFullPath(), 'list' => true, 'view' => true]),
```
Comment (fashxp) [inline tests/Model/Permissions/ModelAssetPermissionsTest.php]: is that related to that PR?

File: tests/Model/Permissions/ModelDocumentPermissionsTest.php [hunk 8]
```diff
@@ -728,23 +728,23 @@ public function testManyElementSearch()
         //update role
         $role = User\Role::getByName('Testrole');
         $role->setWorkspacesDocument([
-            (new User\Workspace\Document())->setValues(['cId' => $manyElementX->getId(), 'cPath' => $manyElementX->getFullpath(), 'list' => true, 'view' => true]),
-            (new User\Workspace\Document())->setValues(['cId' => $this->groupfolder->getId(), 'cPath' => $this->groupfolder->getFullpath(), 'list' => true, 'view' => true]),
+            (new User\Workspace\Document())->setValues(['cId' => $manyElementX->getId(), 'cPath' => $manyElementX->getRealFullPath(), 'list' => true, 'view' => true]),
```
Comment (fashxp) [inline tests/Model/Permissions/ModelDocumentPermissionsTest.php]: is that related to that PR?

File: tests/_support/Test/EcommerceTestCase.php [hunk 9]
```diff
@@ -49,12 +50,12 @@ protected function buildEnvironment(): EnvironmentInterface
     protected function buildSession(): SessionInterface
     {
         if (null === $this->session) {
-            $this->session = new Session(new MockArraySessionStorage());
+            $this->session = \Pimcore::getContainer()->get('session');
```
Comment (fashxp) [inline tests/_support/Test/EcommerceTestCase.php]: is there a session service in the future?

File: bundles/EcommerceFrameworkBundle/PricingManager/PricingManager.php [hunk 10]
```diff
@@ -228,7 +231,7 @@ public function getValidRules()
     public function getEnvironment()
     {
         /** @var AttributeBagInterface $sessionBag */
-        $sessionBag = $this->session->getBag(SessionConfigurator::ATTRIBUTE_BAG_PRICING_ENVIRONMENT);
+        $sessionBag = $this->session->getBag(SessionBagListener::ATTRIBUTE_BAG_PRICING_ENVIRONMENT);
 
         $class = $this->options['environment_class'];
```
Comment (fashxp) [inline bundles/EcommerceFrameworkBundle/PricingManager/PricingManager.php:237]: we also should add deprecation notices to the `get/setSession` from `Pimcore\Bundle\EcommerceFrameworkBundle\PricingManager\EnvironmentInterface`

File: bundles/EcommerceFrameworkBundle/CartManager/SessionCart.php [hunk 11]
```diff
@@ -44,7 +44,7 @@ protected function getCartCheckoutDataClassName()
     protected static function getSessionBag(): AttributeBagInterface
     {
         /** @var AttributeBagInterface $sessionBag */
-        $sessionBag = \Pimcore::getContainer()->get('session')->getBag(SessionConfigurator::ATTRIBUTE_BAG_CART);
+        $sessionBag = \Pimcore::getContainer()->get('request_stack')->getSession()->getBag(SessionBagListener::ATTRIBUTE_BAG_CART);
```
Comment (fashxp) [inline bundles/EcommerceFrameworkBundle/CartManager/SessionCart.php]: do we need a fallback there if there is no `request_stack` available? at least for BC reasons?

**Review by fashxp — CHANGES_REQUESTED** [review]

File: bundles/EcommerceFrameworkBundle/Resources/config/services.yaml (hunk 4 above)
Comment (dvesh3) [reply bundles/EcommerceFrameworkBundle/Resources/config/services.yaml]: `SessionConfigurator` extends from new `SessionBagListener`. Tags are removed in the core implementations otherwise they will trigger the deprecation warning here https://github.com/pimcore/pimcore/pull/12294/files#diff-461ad6908b3c6404bf8f5e543677e341e288d786e9038e8af7998fe82285026aR69-R72

File: tests/Model/Permissions/ModelDocumentPermissionsTest.php (hunk 8 above)
Comment (dvesh3) [reply tests/Model/Permissions/ModelDocumentPermissionsTest.php]: It fixes the tests

File: bundles/EcommerceFrameworkBundle/Tools/SessionConfigurator.php (hunk 6 above)
Comment (dvesh3) [reply bundles/EcommerceFrameworkBundle/Tools/SessionConfigurator.php:19]: Should be still accessible through `SessionConfigurator::ATTRIBUTE_BAG_CART` as `SessionConfigurator` extends `SessionBagListener`. `clearSession` is still there in `SessionBagListener`.

File: bundles/EcommerceFrameworkBundle/PricingManager/Condition/Token.php (hunk 3 above)
Comment (dvesh3) [reply bundles/EcommerceFrameworkBundle/PricingManager/Condition/Token.php:23]: done

File: lib/Targeting/Session/SessionConfigurator.php (hunk 5 above)
Comment (dvesh3) [reply lib/Targeting/Session/SessionConfigurator.php]: what do you mean by move? these methods should be there since `SessionConfigurator` extends `TargetingSessionBagListener`

File: tests/Model/Permissions/ModelAssetPermissionsTest.php (hunk 7 above)
Comment (dvesh3) [reply tests/Model/Permissions/ModelAssetPermissionsTest.php]: It fixes the tests

File: tests/_support/Test/EcommerceTestCase.php (hunk 9 above)
Comment (dvesh3) [reply tests/_support/Test/EcommerceTestCase.php]: you are right. Reverted the change.

File: bundles/EcommerceFrameworkBundle/PricingManager/PricingManager.php (hunk 10 above)
Comment (dvesh3) [reply bundles/EcommerceFrameworkBundle/PricingManager/PricingManager.php:237]: done

File: lib/Targeting/Session/SessionConfigurator.php (hunk 5 above)
Comment (fashxp) [reply lib/Targeting/Session/SessionConfigurator.php]: uh, I missed the parent class. 
but `SessionConfiguratorInterface` we should not remove, should we?

File: bundles/EcommerceFrameworkBundle/Tools/SessionConfigurator.php (hunk 6 above)
Comment (fashxp) [reply bundles/EcommerceFrameworkBundle/Tools/SessionConfigurator.php:19]: uh, I missed the parent class.
but the interface `SessionConfiguratorInterface` we should still keep?

File: tests/_support/Test/EcommerceTestCase.php (hunk 9 above)
Comment (fashxp) [reply tests/_support/Test/EcommerceTestCase.php]: revert not done yet.

**Review by fashxp — CHANGES_REQUESTED** [review]

File: tests/_support/Test/EcommerceTestCase.php (hunk 9 above)
Comment (dvesh3) [reply tests/_support/Test/EcommerceTestCase.php]: done

File: lib/Targeting/Session/SessionConfigurator.php (hunk 5 above)
Comment (dvesh3) [reply lib/Targeting/Session/SessionConfigurator.php]: Implemented `SessionConfiguratorInterface` in deprecated class

File: bundles/EcommerceFrameworkBundle/Tools/SessionConfigurator.php (hunk 6 above)
Comment (dvesh3) [reply bundles/EcommerceFrameworkBundle/Tools/SessionConfigurator.php:19]: Implemented `SessionConfiguratorInterface` in deprecated class

File: bundles/EcommerceFrameworkBundle/CartManager/SessionCart.php (hunk 11 above)
Comment (dvesh3) [reply bundles/EcommerceFrameworkBundle/CartManager/SessionCart.php]: done here https://github.com/pimcore/pimcore/pull/12241/files#diff-5800088a556ed69ab8bbc37c87de139807255039d98f4b6c1b18f01cdea221d3

**Comment by dvesh3** [comment]
This should wait for https://github.com/pimcore/pimcore/pull/12241 merge and rebase for tests to pass.

File: bundles/EcommerceFrameworkBundle/PricingManager/Environment.php [hunk 12]
```diff
@@ -212,20 +212,36 @@ public function getCategories()
...
     /**
+     * @deprecated
+     *
      * @param AttributeBagInterface $session
      *
      * @return EnvironmentInterface
      */
     public function setSession(AttributeBagInterface $session)
     {
+        trigger_deprecation(
```
Comment (fashxp) [inline bundles/EcommerceFrameworkBundle/PricingManager/Environment.php]: remove that message as it would spam the logs

File: bundles/EcommerceFrameworkBundle/PricingManager/Environment.php [hunk 13]
```diff
@@ -212,20 +212,36 @@ public function getCategories()
...
+     * @deprecated
+     *
      * @return AttributeBagInterface|null
      */
     public function getSession()
     {
+        trigger_deprecation(
+            'pimcore/pimcore',
+            '10.5',
+            sprintf('%s is deprecated. It will be removed in Pimcore 11.', __METHOD__)
```
Comment (fashxp) [inline bundles/EcommerceFrameworkBundle/PricingManager/Environment.php]: add info that getting AND setting session is deprecated

**Review by fashxp — CHANGES_REQUESTED** [review]

File: bundles/EcommerceFrameworkBundle/PricingManager/Environment.php (hunk 12 above)
Comment (dvesh3) [reply bundles/EcommerceFrameworkBundle/PricingManager/Environment.php]: Done

File: bundles/EcommerceFrameworkBundle/PricingManager/Environment.php (hunk 13 above)
Comment (dvesh3) [reply bundles/EcommerceFrameworkBundle/PricingManager/Environment.php]: Done

**Review by fashxp — APPROVED** [review]
